### PR TITLE
Wallet state transition cleanup

### LINKED ISSUE #255: Clean up wallet state validation and state transitions
We need to go through all functions in `Bridge` libraries and `Wallet` library and see if there is a way to align them in terms of where the wallet state is validated. Also, we should clean up the naming of functions performing state transitions.

### PR DESCRIPTION
**Pull request by pdyraga**
Closes #255

#333 cleans up where the wallet state is validated for functions not mutating the state. The rule is that the state must be validated in the function handling the given request, usually in the very first line.

This PR cleans up state validation for functions that update the wallet state. The rule is:
Only `Wallet` library can move the wallet between states and the input state validation must be done inside `Wallets` library.

Bridge function | Allowed wallet states | Wallet state validation | Transitioning to state
-- | -- | -- | --
Bridge.notifyRedemptionTimeout | Live, MovingFunds, Terminated | Wallets.notifyWalletRedemptionTimeout | MovingFunds (only if Live) or Closing (only if Live if no main UTXO set)
Bridge.submitMovingFundsProof | MovingFunds | Wallets.notifyWalletFundsMoved | Closing
Bridge.notifyMovingFundsTimeout | MovingFunds | Wallets.notifyWalletMovingFundsTimeout | Terminated
Bridge.notifyMovingFundsBelowDust | MovingFunds | Wallets.notifyWalletMovingFundsBelowDust | Closing
Bridge.notifyMovedFundsSweepTimeout | Live, MovingFunds, Terminated | Wallets.notifyWalletMovedFundsSweepTimeout | Terminated (only if Live or MovingFunds)
Bridge.__ecdsaWalletCreatedCallback | Unknown | Wallet.registerNewWallet | Live
Bridge.__ecdsaWalletHeartbeatFailedCallback | Live | Wallet.notifyWalletHeartbeatFailed | MovingFunds or Closing if no main UTXO set
Bridge.notifyWalletCloseable | Live | Wallet.notifyWalletCloseable | MovingFundsor Closing if no main UTXO set
Bridge.notifyWalletClosingPeriodElapsed | Closing | Wallet.notifyWalletClosingPeriodElapsed | Closed
Bridge.notifyFraudChallengeDefeatTimeout | Live, MovingFunds, Closing, Terminated | Wallets.notifyWalletFraudChallengeDefeatTimeout | Terminated

I recommend reviewing this PR commit-by-commit. The gas cost difference is negligible - sometimes it improved a bit, and sometimes it is getting worse a bit.

### CODE REVIEW

**Review by tomaszslabon — DISMISSED** [review]

**Comment by pdyraga** [comment]
@tomaszslabon I've pushed one change updating docs for `Terminated` state, please look again.

File: solidity/contracts/bridge/Wallets.sol [hunk 1]
```diff
@@ -569,4 +569,33 @@ library Wallets {
 
         beginWalletClosing(self, walletPubKeyHash);
     }
+
+    /// @notice Called when the timeout for MovingFunds wallet elapsed.
+    ///         Slashes wallet members and terminates the wallet.
+    /// @dev Requirements:
+    ///      - The wallet must be in the MovingFunds state.
```
Comment (lukasz-zimnoch) [inline solidity/contracts/bridge/Wallets.sol]: Let's document the parameters as well.

File: solidity/contracts/bridge/Wallets.sol [hunk 2]
```diff
@@ -598,4 +598,24 @@ library Wallets {
 
         terminateWallet(self, walletPubKeyHash);
     }
+
+    /// @notice Called when a MovingFunds wallet has a balance below the dust
+    ///         threshold. Begins the wallet closing.
+    /// @dev Requirements:
+    ///      - The wallet must be in the MovingFunds state.
```
Comment (lukasz-zimnoch) [inline solidity/contracts/bridge/Wallets.sol]: Let's document the parameters.

File: solidity/contracts/bridge/Wallets.sol (hunk 1 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/Wallets.sol]: I added those comments from the commit view so this probably caused confusion. You are right, in the final version, this is the only occurrence.

**Review by lukasz-zimnoch — APPROVED** [review]

**Pull request merged**